src/SprykerEco/Zed/AvalaraTaxShipmentTax/Business/Calculator/AbstractShipmentAvalaraTaxCalculator.php

```diff
@@ -0,0 +1,162 @@
+        AvalaraCreateTransactionResponseTransfer $avalaraCreateTransactionResponseTransfer
+    ): CalculableObjectTransfer;
+
+    /**
+     * @param \ArrayObject|\Generated\Shared\Transfer\AvalaraTransactionLineTransfer[] $avalaraTransactionLineTransfers
+     *
+     * @return \Generated\Shared\Transfer\AvalaraTransactionLineTransfer[][]
+     */
+    protected function getProductOptionAvalaraTransactionLineTransfersIndexedByItemSkuAndProductOptionSku(ArrayObject $avalaraTransactionLineTransfers): array
+    {
+        $mappedProductOptionAvalaraTransactionLineTransfers = [];
+        foreach ($avalaraTransactionLineTransfers as $avalaraTransactionLineTransfer) {
+            if ($avalaraTransactionLineTransfer->getRef1OrFail() !== AvalaraLineItemMapper::SHIPMENT_AVALARA_LINE_TYPE) {
+                continue;
+            }
+
+            $itemSku = $avalaraTransactionLineTransfer->getRef2OrFail();
+            $productOptionSku = $avalaraTransactionLineTransfer->getItemCodeOrFail();
+            $mappedProductOptionAvalaraTransactionLineTransfers[$itemSku][$productOptionSku] = $avalaraTransactionLineTransfer;
+        }
+
+        return $mappedProductOptionAvalaraTransactionLineTransfers;
+    }
+
+    /**
+     * @param string $transactionLineDetails
+     *
+     * @return float
+     */
+    protected function sumTaxRateFromTransactionLineDetails(string $transactionLineDetails): float
+    {
+        $taxRateSum = 0.0;
+
+        /** @var \Avalara\TransactionLineDetailModel[] $transactionLineDetailModels */
+        $transactionLineDetailModels = $this->utilEncodingService->decodeJson($transactionLineDetails, false);
+        foreach ($transactionLineDetailModels as $transactionLineDetailModel) {
+            $taxRateSum += $transactionLineDetailModel->rate ?? 0.0;
+        }
+
+        return $this->convertToPercents($taxRateSum);
+    }
+
+    /**
+     * @param float $number
+     *
+     * @return float
+     */
+    protected function convertToPercents(float $number): float
+    {
+        return $number * 100.0;
+    }
+
+    /**
+     * @param \ArrayObject|\Generated\Shared\Transfer\ExpenseTransfer[] $expenseTransfers
+     * @param \Generated\Shared\Transfer\ShipmentTransfer $shipmentTransfer
+     *
+     * @return \Generated\Shared\Transfer\ExpenseTransfer|null
+     */
+    protected function findQuoteExpenseByShipment(
+        ArrayObject $expenseTransfers,
+        ShipmentTransfer $shipmentTransfer
+    ): ?ExpenseTransfer {
+        $itemShipmentKey = $this->shipmentService->getShipmentHashKey($shipmentTransfer);
+        foreach ($expenseTransfers as $expenseTransfer) {
+            if (!$expenseTransfer->getShipment()) {
+                continue;
+            }
+
+            $expenseShipmentKey = $this->shipmentService->getShipmentHashKey($expenseTransfer->getShipmentOrFail());
+
+            if ($expenseShipmentKey === $itemShipmentKey && $expenseTransfer->getType() === static::SHIPMENT_EXPENSE_TYPE) {
```

Comment:
* We can do and an early return `if($expenseTransfer->getType() !== static::SHIPMENT_EXPENSE_TYPE)continue;`.

Reply:
Resolved.